src/default-directory-searcher.coffee

```diff
@@ -0,0 +1,61 @@
+  #
+  # Returns a `boolean` indicating whether this object can search this `Directory`.
+  canSearchDirectory: (directory) -> true
+
+  # Public: Performs a text search for files in the specified `Directory`, subject to the
+  # specified parameters.
+  #
+  # Results are streamed back to the caller via `recordSearchResult()` and `recordSearchError()`.
+  #
+  # * `directory` {Directory} that has been accepted by this provider's `canSearchDirectory()`
+  # predicate.
+  # * `regexSource` {String} regex to search with. Produced via `RegExp::source`.
+  # (Note this reflects the "Use Regex" option exposed via the ProjectFindView UI.)
+  # * `onSearchResult` {Function} Should be called with each matching search result.
+  #   * `searchResult` {Object} with the following keys:
+  #     * `filePath` {String} absolute path to the matching file.
+  #     * `matches` {Array} with object elements with the following keys:
+  #       * `lineText` {String} The full text of the matching line (without a line terminator character).
+  #       * `lineTextOffset` {Number} (This always seems to be 0?)
+  #       * `matchText` {String} The text that matched the `regex` used for the search.
+  #       * `range` {Range} Identifies the matching region in the file. (Likely as an array of numeric arrays.)
+  # * `onSearchError` {Function} Should be called to report a search error.
+  # * `onPathsSearched` {Function} callback that should be invoked periodically with the number of
+  # paths searched.
+  # * `options` {Object} with the following properties:
+  #   * `ignoreCase` {boolean}
+  #   * `inclusions` {Array} of glob patterns (as strings) to search within. Note that this
+  #   array may be empty, indicating that all files should be searched.
+  #
+  #   Each item in the array is a file/directory pattern, e.g., `src` to search in the "src"
+  #   directory or `*.js` to search all JavaScript files. In practice, this often comes from the
+  #   comma-delimited list of patterns in the bottom text input of the ProjectFindView dialog.
+  #   * `ignoreHidden` {boolean}
+  #   * `excludeVcsIgnores` {boolean}
+  #   * `exclusions` {Array} similar to inclusions
+  #   * `follow` {boolean} whether symlinks should be followed
+  #
+  # Returns a `Promise` that includes a `cancel()` method. If invoked before the `Proimse` is
+  # determined, it will reject the `Promise`.
+  search: (directory, regexSource, onSearchResult, onSearchError, onPathsSearched, options) ->
+    task = null
+    rootPaths = [directory.getPath()]
+    promise = new Promise (resolve, reject) ->
+      task = Task.once require.resolve('./scan-handler'), rootPaths, regexSource, options, resolve
+      task.on 'task:cancelled', reject
+    promise.cancel = ->
+      task.cancel()
+
+    task.on 'scan:result-found', onSearchResult
+    task.on 'scan:file-error', onSearchError
+    task.on 'scan:paths-searched', onPathsSearched
+
+    promise
```

Comment:
> At a minimum, I think it would have to continue to be a thenable such that is "Promise compatible,"

That seems fine to me. I just think that since the returned object needs to have a `.cancel()` method, it would be clearer if were an instance of our own class, not of `Promise`.

> Then the caller passes in an instance of DirectorySearchDelegate, which avoids a bunch of event registration.

I see your point about the smaller amount of indirection involved when using the delegate pattern (one callback instead of potentially n callbacks), but in our situation I don't think there is much overhead or complexity involved in using event registration. All of Atom's existing APIs use the observer pattern, and it fits in well with our other tools like `CompositeDisposable`.

Also, I think it would be simpler implementation-wise as well, at least in the case of the `DefaultDirectorySearcher`. Rather than having to define a callback, like this: